# Data preprocessing

Man Ho Wong | March 21, 2022


In [19]:
# Python libraries and packages required
from collections import Counter
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pickle
import pprint                    # For pretty printing
import pylangacq                 # For reading CHAT files
from tqdm import tqdm            # For showing progress bar

# use ggplot style for figures
plt.style.use('ggplot')

# Unpickle data
data = pickle.load(open('../data/childes/corpus_info.pkl', 'rb'))
data_idx = data[1]

# Pretty printing for better readability
# With this, you can print dict in compact format instead of one item
#   per line. Items will be in alphabetical order. Nested Dict will be printed 
#   with suitable indentation. It will also print Counter in descending order.
#   To use this, call cp.pprint()
cp = pprint.PrettyPrinter(compact=True, sort_dicts=True)

# Function to get all labels of a given variable (e.g. 'mot_edu')
def get_labels(var):
    labels_by_corpus = {}
    corpus_set = set(data_idx.corpus)
    for c in corpus_set:
        labels_by_corpus[c] = set(data_idx[var][data_idx.corpus==c])
    return labels_by_corpus

`data_idx` is a `DataFrame` created by [data_curation.ipynb](Data_Science/Child-Vocab-Development/codes/data_curation.ipynb). It contains basic information about the files in the curated dataset. 

In [2]:
data_idx

file_path     corpus  year  \
0                  ../data/childes/Bates/Free20/amy.cha      Bates   NaN   
1                ../data/childes/Bates/Free20/betty.cha      Bates   NaN   
2                ../data/childes/Bates/Free20/chuck.cha      Bates   NaN   
3                 ../data/childes/Bates/Free20/doug.cha      Bates   NaN   
4                   ../data/childes/Bates/Free20/ed.cha      Bates   NaN   
...                                                 ...        ...   ...   
2606  ../data/childes/VanHouten/Twos/teaching/parkt.cha  VanHouten   NaN   
2607  ../data/childes/VanHouten/Twos/teaching/pricet...  VanHouten   NaN   
2608  ../data/childes/VanHouten/Twos/teaching/raidt.cha  VanHouten   NaN   
2609  ../data/childes/VanHouten/Twos/teaching/riott.cha  VanHouten   NaN   
2610  ../data/childes/VanHouten/Twos/teaching/royalt...  VanHouten   NaN   

     participants          name  age_d  age_m     sex           group  \
0      {MOT, CHI}  Target_Child    600   20.0  female              TD   
1      {MOT, CHI}         Betty    600   20.0  female              TD   
2      {MOT, CHI}         Chuck    600   20.0    male              TD   
3      {MOT, CHI}          Doug    600   20.0    male              TD   
4      {MOT, CHI}            Ed    600   20.0    male              TD   
...           ...           ...    ...    ...     ...             ...   
2606   {MOT, CHI}       Matthew    840   28.0    male       MOT_Older   
2607   {MOT, CHI}         Peter    840   28.0    male  MOT_adolescent   
2608   {MOT, CHI}         Tommy    840   28.0    male       MOT_older   
2609   {MOT, CHI}        Robert    840   28.0    male  MOT_Adolescent   
2610   {MOT, CHI}         Sarah    840   28.0  female  MOT_Adolescent   

              ses      mot_edu  \
0              MC  unspecified   
1              MC  unspecified   
2              MC  unspecified   
3              MC  unspecified   
4              MC  unspecified   
...           ...          ...   
2606  unspecified  unspecified   
2607  unspecified  unspecified   
2608  unspecified  unspecified   
2609  unspecified  unspecified   
2610  unspecified  unspecified   

                                              situation   activities  \
0                                           unspecified  unspecified   
1                                           unspecified  unspecified   
2                                           unspecified  unspecified   
3                                           unspecified  unspecified   
4                                           unspecified  unspecified   
...                                                 ...          ...   
2606  Teaching. Mother and Child sit on floor. First...  unspecified   
2607  Teaching. Mother and Child sit on the floor in...  unspecified   
2608  Teaching. Mother , Child and Father sit in liv...  unspecified   
2609  Teaching. Mother and Child sit on the living r...  unspecified   
2610  Teaching. Mother and Child sit on living room ...  unspecified   

              study_type  
0     cross, toyplay, TD  
1     cross, toyplay, TD  
2     cross, toyplay, TD  
3     cross, toyplay, TD  
4     cross, toyplay, TD  
...                  ...  
2606  cross, toyplay, TD  
2607  cross, toyplay, TD  
2608  cross, toyplay, TD  
2609  cross, toyplay, TD  
2610  cross, toyplay, TD  

[2611 rows x 14 columns]

# Data cleaning and integration

As mentioned in `data_curation.ipynb`, not all corpora use the same labels for some variables. For example, some corpora use school grade year to define mother's education (`mot_edu`), and some use level of education (e.g. 'college'):

In [3]:
mot_edu_labels = set(data_idx.mot_edu)
cp.pprint(mot_edu_labels)  # Print compactly

{'**', '10', '10 , GED', '102', '11', '11 , GED', '11+', '12', '12+', '13',
 '13+', '14', '15', '16', '6', '7', '8', '9', "Associate's_Degree", 'College',
 'College_Doctoral', 'College_J.D.', "College_Master's", 'High_School_Diploma',
 'MOT_1', 'MOT_2', 'MOT_3', 'Some_College', 'XX', 'almost 12', 'unspecified'}


Besides `mot_edu`, other variables with different labels used in different corpora include:  
- `group` (child's developmental group, e.g. typically developed)
- `ses` (Child's SES (socioeconomic status; same as mother's))
- `situation` Situation of recording (e.g. play session)
- `activities` (Activities during recording, e.g. Toy play)
- `study_type` type of study (e.g. longitudinal study)
    
I will change the labels for some of these variables, so that all corpora use the same set of labels.

Below is a list of homepages for all the corpora appeared in the curated dataset. You can find the basic information about the corpus, such as variable and label definitions, on its homepage and in the linked publications. I will check these pages when the variable labels are ambiguous.

In [4]:
corpus_homepages = {}
corpus_set = set(data_idx.corpus)

for c in corpus_set:
    url = "https://childes.talkbank.org/access/Eng-NA/" + c + ".html"
    corpus_homepages[c] = url

cp.pprint(corpus_homepages)

{'Bates': 'https://childes.talkbank.org/access/Eng-NA/Bates.html',
 'Bernstein': 'https://childes.talkbank.org/access/Eng-NA/Bernstein.html',
 'Brown': 'https://childes.talkbank.org/access/Eng-NA/Brown.html',
 'Clark': 'https://childes.talkbank.org/access/Eng-NA/Clark.html',
 'Demetras2': 'https://childes.talkbank.org/access/Eng-NA/Demetras2.html',
 'Gleason': 'https://childes.talkbank.org/access/Eng-NA/Gleason.html',
 'HSLLD': 'https://childes.talkbank.org/access/Eng-NA/HSLLD.html',
 'Hall': 'https://childes.talkbank.org/access/Eng-NA/Hall.html',
 'Hicks': 'https://childes.talkbank.org/access/Eng-NA/Hicks.html',
 'Nelson': 'https://childes.talkbank.org/access/Eng-NA/Nelson.html',
 'NewmanRatner': 'https://childes.talkbank.org/access/Eng-NA/NewmanRatner.html',
 'Post': 'https://childes.talkbank.org/access/Eng-NA/Post.html',
 'VanHouten': 'https://childes.talkbank.org/access/Eng-NA/VanHouten.html'}


### `mot_edu`

I will start with the 'simplest' variable, `mot_edu`.
First, let's find out which of the above labels each corpus uses for `mot_edu`. I will use a self-defined function, `get_labels()` (see header section of this notebook):

In [5]:
cp.pprint(get_labels('mot_edu'))  # get_labels() is defined in header section

{'Bates': {'unspecified'},
 'Bernstein': {'unspecified'},
 'Brown': {'unspecified'},
 'Clark': {'unspecified'},
 'Demetras2': {'unspecified'},
 'Gleason': {'unspecified'},
 'HSLLD': {'**', '10', '10 , GED', '102', '11', '11 , GED', '11+', '12', '12+',
           '13', '13+', '14', '15', '16', '6', '7', '8', '9', 'XX', 'almost 12',
           'unspecified'},
 'Hall': {'unspecified'},
 'Hicks': {'unspecified'},
 'Nelson': {'unspecified'},
 'NewmanRatner': {"Associate's_Degree", 'College', 'College_Doctoral',
                  'College_J.D.', "College_Master's", 'High_School_Diploma',
                  'Some_College', 'unspecified'},
 'Post': {'unspecified'},
 'VanHouten': {'unspecified', 'MOT_3', 'MOT_2', 'MOT_1'}}


--
  
As shown here, different classifications of education level are used in three corpora (HSLLD, NewmanRatner and VanHouten). HSLLD has a very detailed classification while VanHouten has only three classes. To merge all the labels across different corpora, I will use a less detailed classification without losing information needed for this project.

The labels used in HSLLD correspond to the school grade year the mother achieved. In the U.S. educational system, 'GED' is equivalent to completing grade 12, so I will merge these two classes as one. The labels '13' and beyond correspond to some post-secondary education (e.g. vocational training).

According to VanHouten's documentation, mother’s education is classified as MOT_1 (completed junior high), MOT_2 (completed high school) or MOT_3 (some post-secondary education).

I will merge the current labels into the following labels according to their common definitions: 

- `JH-`, `HS-`, `HS`, `HS+` for:
    - before completing junior high (before grade 9),
    - before completing high school (before GED or grade 12),
    - completed high school (GED or grade 12), and
    - some post-secondary education (after high school).

- `AD`, `UG-`, `UG` for:
    - associate's degree,
    - some undergraduate education, and
    - bachelor's degree.

- `MS`, `JD`, `DR` for:
    - master's degree,
    - juris doctor degree, and
    - doctoral degree.
    
I will map the current `mot_edu` labels to the above new labels with a `dictionary` and update `data_idx` with the new labels:

In [6]:
# Function to update 'mot_edu'
def update_mot_edu(data):
    label = data
    #' mot_edu' label mapping
    EDU_DICT = {
              'JH-':['6','7','8'],
              'HS-':['9','10','11','11+','almost 12','MOT_1'],
              'HS':['10 , GED','11 , GED','12','12+',
                    'High_School_Diploma','MOT_2'],
              'HS+':['13','13+','14','15','16','MOT_3'],
              'AD':["Associate's_Degree"],
              'UG-':['Some_College'],
              'UG':['College'], 
              'MS':["College_Master's"],
              'JD':['College_J.D.'],    
              'DR':['College_Doctoral'],
              'unspecified':['**','102','XX','unspecified']
               }
    
    for key in EDU_DICT:
        if data in EDU_DICT[key]:
            label = key
    return label

# Update 'mot_edu'
data_idx.mot_edu = data_idx.mot_edu.map(update_mot_edu)

# Check if update was successful
cp.pprint(get_labels('mot_edu'))

{'Bates': {'unspecified'},
 'Bernstein': {'unspecified'},
 'Brown': {'unspecified'},
 'Clark': {'unspecified'},
 'Demetras2': {'unspecified'},
 'Gleason': {'unspecified'},
 'HSLLD': {'unspecified', 'HS', 'HS+', 'JH-', 'HS-'},
 'Hall': {'unspecified'},
 'Hicks': {'unspecified'},
 'Nelson': {'unspecified'},
 'NewmanRatner': {'JD', 'unspecified', 'HS', 'UG', 'UG-', 'DR', 'MS', 'AD'},
 'Post': {'unspecified'},
 'VanHouten': {'HS-', 'unspecified', 'HS', 'HS+'}}


--

All labels for `mot_edu` were updated sucessfully!

### `group`

Below are the labels used in different corpora:

In [7]:
cp.pprint(get_labels('group'))

{'Bates': {'TD'},
 'Bernstein': {'TD', 'unspecified'},
 'Brown': {'TD', 'unspecified'},
 'Clark': {'TD'},
 'Demetras2': {'TD', 'unspecified'},
 'Gleason': {'TD', 'normal', 'unspecified', 'typical'},
 'HSLLD': {'unspecified'},
 'Hall': {'White,UC', 'TD', 'unspecified'},
 'Hicks': {'unspecified'},
 'Nelson': {'unspecified'},
 'NewmanRatner': {'TD'},
 'Post': {'TD'},
 'VanHouten': {'MOT_Adolescent', 'MOT_Adolescent_', 'MOT_Older', 'MOT_Older_',
               'MOT_adolescent', 'MOT_older', 'TD', 'unspecified'}}


--

As shown above, the `group` information of each file is not properly speficied in every corpus: 
- According the documentations of the corpora, all children in the dataset should be typically developed (`TD` in `group`). However, not all corpora put this information in the `group` field in the header of each file. Therefore, I will change the `unspecified` label to `TD`.
- in the Gleason corpus, labels `normal` and `typical` are used in additon to `TD`. I will change all labels to `TD`.
- In the Hall corpus, the labels for `group` and `ses` were switched in some files (e.g. `White,UC` should be a label for `ses`). I will switch them first and change all `group` labels to `TD`.
- In the VanHouten corpus, all children are `TD`, but it also contains data from adolescent mothers which I will not use for this project. I will first remove these files and change all labels to `TD`.

Let's take a quick look at the Hall corpus first:

In [20]:
labels = ['TD', 'typical', 'normal']
data_idx[['group','ses','study_type']][(data_idx.corpus=='Hall') & 
                                       ~(data_idx.group.isin(labels))]

group          ses            study_type
1471  unspecified     Black,WC  cross, everyday, AAE
1472  unspecified     Black,WC  cross, everyday, AAE
1473  unspecified     Black,WC  cross, everyday, AAE
1474  unspecified     Black,WC  cross, everyday, AAE
1475  unspecified     Black,WC  cross, everyday, AAE
1476  unspecified     Black,WC  cross, everyday, AAE
1477  unspecified     Black,WC  cross, everyday, AAE
1478  unspecified     Black,WC  cross, everyday, AAE
1479  unspecified     Black,WC  cross, everyday, AAE
1480  unspecified     Black,WC  cross, everyday, AAE
1482     White,UC  unspecified   cross, everyday, TD
1485     White,UC  unspecified   cross, everyday, TD
1492  unspecified  unspecified           unspecified

As shown above in row 1482 and 1485, the label `White,UC` (i.e. upper class white) is actually a label for `ses`. Therefore, I will first copy the labels to `ses`, and then change all the `unspecified` labels in `group` to `TD`:

In [21]:
# idx = data_idx.group[data_idx.group=='White,UC'].index
data_idx.loc[data_idx.group=='White,UC','ses'] = 'White,UC'
data_idx.loc[data_idx.corpus=='Hall','group'] = 'TD' 

data_idx[['group','ses','study_type']][(data_idx.corpus=='Hall')]

group          ses            study_type
1460    TD     Black,UC  cross, everyday, AAE
1461    TD     Black,UC  cross, everyday, AAE
1462    TD     Black,UC  cross, everyday, AAE
1463    TD     Black,UC  cross, everyday, AAE
1464    TD     Black,UC  cross, everyday, AAE
1465    TD     Black,UC  cross, everyday, AAE
1466    TD     Black,UC  cross, everyday, AAE
1467    TD     Black,UC  cross, everyday, AAE
1468    TD     Black,UC  cross, everyday, AAE
1469    TD     Black,UC  cross, everyday, AAE
1470    TD     Black,UC  cross, everyday, AAE
1471    TD     Black,WC  cross, everyday, AAE
1472    TD     Black,WC  cross, everyday, AAE
1473    TD     Black,WC  cross, everyday, AAE
1474    TD     Black,WC  cross, everyday, AAE
1475    TD     Black,WC  cross, everyday, AAE
1476    TD     Black,WC  cross, everyday, AAE
1477    TD     Black,WC  cross, everyday, AAE
1478    TD     Black,WC  cross, everyday, AAE
1479    TD     Black,WC  cross, everyday, AAE
1480    TD     Black,WC  cross, everyday, AAE
1481    TD     White,UC   cross, everyday, TD
1482    TD     White,UC   cross, everyday, TD
1483    TD     White,UC   cross, everyday, TD
1484    TD     White,UC   cross, everyday, TD
1485    TD     White,UC   cross, everyday, TD
1486    TD     White,UC   cross, everyday, TD
1487    TD     White,UC   cross, everyday, TD
1488    TD     White,UC   cross, everyday, TD
1489    TD     White,UC   cross, everyday, TD
1490    TD     White,UC   cross, everyday, TD
1491    TD     White,WC   cross, everyday, TD
1492    TD  unspecified           unspecified
1493    TD     White,WC   cross, everyday, TD
1494    TD     White,WC   cross, everyday, TD
1495    TD     White,WC   cross, everyday, TD
1496    TD     White,WC   cross, everyday, TD
1497    TD     White,WC   cross, everyday, TD
1498    TD     White,WC   cross, everyday, TD
1499    TD     White,WC   cross, everyday, TD

### `situation`

In [ ]:
# cp.pprint(get_labels('activities'))

# for corpus in corpus_set:
#     for labels in get_labels('activities')[corpus]:
#         if 'reading' in labels:
#             print(corpus)
            
keywords = []            
for f in data_idx.situation:
    keywords.extend(f.lower().split())

Counter(keywords).most_common()

In [ ]:
f = data_idx.file_path[data_idx.corpus=='Bernstein']
reader = pylangacq.Reader.from_files(f)

In [ ]:
group_df = pd.DataFrame()
group_df['name'] = [reader.headers()[file]['Participants']['CHI']['name'] for file in range(len(f))]
group_df['group'] = [reader.headers()[file]['Participants']['CHI']['group'] for file in range(len(f))]
group_df['participants'] = [reader.participants(by_files=True)[file] for file in range(len(f))]
group_df['mot_edu'] = [reader.headers()[file]['Participants']['CHI']['education'] for file in range(len(f))]
group_df['situation'] = [reader.headers()[file]['Situation'] for file in range(len(f))]
group_df['type'] = [reader.headers()[file]['Types'] for file in range(len(f))]

group_df

# reader.tokens(participants='CHI', by_files=True)[35]

### `ses`

In [ ]:
cp.pprint(get_labels('ses'))

In [ ]:
reader[3].headers()

In [ ]:
reader[0].headers()[0]['Situation'] if reader[0].headers()[0]['Situation'] else 0

if reader[0].headers()[0]['Situation']: print(reader[0].headers()[0]['Situation'])
